Add Rollup ILM Action (#65633)
elastic/elasticsearch: docs/reference/ilm/actions/ilm-rollup.asciidoc

[role="xpack"]
[[ilm-rollup]]
=== Rollup

Phases allowed: hot, cold.

Aggregates an index's time series data and stores the results in a new read-only
index. For example, you can roll up hourly data into daily or weekly summaries.

> For more information about Rollup, check out the <<rollup-api, Rollup Action Documentation>>

Reviewer: I don't think we need the caps here.
```suggestion
For more information about rollup, see the <<rollup-api, rollup action documentation>>
```

> The name of the rolled up index will be the original index name of the managed index prefixed

Reviewer 2: @talevy How does this work in the context of data streams? Does it work in the context of data streams? What are the index names there?

Reviewer: ILM works on the index level, not directly on data streams. If the index is a backing index for a data stream, the rollup index is a backing index for the same stream. The name of the new rollup index is based on the name of the backing index. For example, `.ds-my-data-stream-2021.01.29-000001` would become `rollup-.ds-my-data-stream-2021.01.29-000001`.

This is similar to how the ILM `shrink` action currently renames backing indices:

https://www.elastic.co/guide/en/elasticsearch/reference/master/data-streams.html#data-streams-generation

I hope that helps. I'll open up a follow-up PR to better clarify this in the docs.

Author: Hey @ruflin and @jrodewig,

this is exactly how it works today, but due to some failure-scenarios we discussed recently, we're going to change this behavior.

In the context of datastreams, we will likely randomize the name of the rollup index to prevent collision and conflict with any past or rogue rollup actions. The belief is that the underlying name of the index should not be of concern to users since they will be added into the datastream and managed in ILM. Do you see any problems with this on your end @ruflin?

For this reason, it might make sense to hold off just a little bit on your docs change PR @jrodewig, but much appreciated!

I've opened #68225 to track it. Will likely change next week

Reviewer: Thanks for the update, @talevy!

Reviewer 2: > The belief is that the underlying name of the index should not be of concern to users since they will be added into the datastream and managed in ILM

If this is the case, it is great. I was under the assumption all indices belong to a data stream must use the `.ds-*` prefix and be hidden. So I expect index names like `.ds-rollup-*` or something similar. This would also ensure there are no conflicts and makes it clear that these rollups belong to the data streams.

The `rollup-` prefix is what put me on the wrong track here and made me assume it does not belong to the data stream.

Author: it is actually common for indices managed by ILM to be renamed with a new prefix. For example, shrunken indices start with the `shrunk-` prefix. 

independent of the name, the indices will be hidden and added to the backing datastream

Reviewer 2: @talevy Interesting. Why are these not prefixed by a `.` and not inside `.ds-*-shrunk-*` or similar to indicate they belong together? I think I'm derailing but could you share a link where I can read up on this? The part I'm concerned is that this not seem to fully fit into our naming scheme: https://www.elastic.co/blog/an-introduction-to-the-elastic-data-stream-naming-scheme

with `rollup-`.

[[ilm-rollup-options]]
==== Rollup options
`config`::
(Required, integer)

> The Rollup configuration, a more detailed description of the
> Rollup configuration specification can be found <<rollup-api-request-body,here>>.

Reviewer: ```suggestion
The rollup configuration, a more detailed description of the
rollup configuration specification can be found <<rollup-api-request-body,here>>.
```

`rollup_policy`::
(Optional, string)

> The name of an  <<index-lifecycle-management, {ilm}>> ({ilm-init}) policy to associate

Reviewer: Removes a duplicate space.
```suggestion
The name of an <<index-lifecycle-management, {ilm}>> ({ilm-init}) policy to associate
```
